backend/layers/processing/dataset_metadata_update.py

```diff
@@ -0,0 +1,288 @@
+            self.artifact_bucket,
+            DatasetStatusKey.RDS,
+            datasets_bucket=self.datasets_bucket,
+        )
+        os.remove(seurat_filename)
+        self.update_processing_status(new_dataset_version_id, DatasetStatusKey.RDS, DatasetConversionStatus.CONVERTED)
+
+    def update_cxg(
+        self,
+        cxg_uri: str,
+        new_cxg_dir: str,
+        dataset_version_id: DatasetVersionId,
+        metadata_update: DatasetArtifactMetadataUpdate,
+    ):
+        self.s3_provider.upload_directory(cxg_uri, new_cxg_dir)
+        ctx = tiledb.Ctx(H5ADDataFile.tile_db_ctx_config)
+        array_name = f"{new_cxg_dir}/cxg_group_metadata"
+        with tiledb.open(array_name, mode="r", ctx=ctx) as metadata_array:
+            cxg_metadata_dict = json.loads(metadata_array.meta["corpora"])
+            cxg_metadata_dict.update(metadata_update.as_dict_without_none_values())
+
+        with tiledb.open(array_name, mode="w", ctx=ctx) as metadata_array:
+            metadata_array.meta["corpora"] = json.dumps(cxg_metadata_dict)
+
+        self.business_logic.add_dataset_artifact(dataset_version_id, DatasetArtifactType.CXG, new_cxg_dir)
+        self.update_processing_status(dataset_version_id, DatasetStatusKey.CXG, DatasetConversionStatus.CONVERTED)
+
+    def update_metadata(
+        self,
+        current_dataset_version_id: DatasetVersionId,
+        new_dataset_version_id: DatasetVersionId,
+        metadata_update: DatasetArtifactMetadataUpdate,
+    ):
+        current_dataset_version = self.business_logic.get_dataset_version(current_dataset_version_id)
+        if current_dataset_version.status.processing_status != DatasetProcessingStatus.SUCCESS:
+            self.logger.info(
+                f"Dataset {current_dataset_version_id} is not successfully processed. Skipping metadata update."
+            )
+            return
+
+        artifact_uris = {artifact.type: artifact.uri for artifact in current_dataset_version.artifacts}
+
+        new_dataset_version = self.business_logic.get_dataset_version(new_dataset_version_id)
+        if new_dataset_version.status.processing_status != DatasetProcessingStatus.INITIALIZED:
+            self.logger.info(
+                f"Dataset {new_dataset_version_id} has processing status "
+                f"{new_dataset_version.status.processing_status} rather than expected INITIALIZED."
+                f"Skipping metadata update."
+            )
+            return
+
+        if DatasetArtifactType.RAW_H5AD in artifact_uris:
+            raw_h5ad_uri = artifact_uris[DatasetArtifactType.RAW_H5AD]
+        else:
+            self.logger.error(f"Cannot find raw H5AD artifact uri for {current_dataset_version_id}.")
+            raise ValueError
+
+        self.process(new_dataset_version_id, raw_h5ad_uri, self.artifact_bucket)
+
+        new_artifact_key_prefix = self.get_key_prefix(new_dataset_version_id.id)
+
+        artifact_jobs = []
+
+        with multiprocessing.Pool() as pool:
```

Comment:
I would process each file in it's own directory. I'm concern there is a change we try uploading the h5ad or seurat file to the CXG directory when we run upload_directory.

Reply:
Understand the concern but this isn't going to be an issue, I don't think. To my eye, the cxg components are all copied s3-to-s3, and then the metadata update is actually applied via tiledb directly (perhaps with local in-mem intermediary under the hood to some degree but that is irrelevant) to the appropriate file in s3. @nayib-jose-gloria can confirm

Reply:
+1 to Dan's comment